### PR TITLE
Mx 15033

### PR DESCRIPTION
**Pull request by dragos-rebegea**
## Reasoning behind the pull request
- process of clearing vote history

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: vm/systemSmartContracts/governance.go [hunk 1]
```diff
@@ -579,7 +582,13 @@ func (g *governanceContract) updateUserVoteListV2(address []byte, nonce uint64,
 		return err
 	}
 
-	return g.saveUserVotesV2(address, userVoteList)
+	err = g.saveUserVotesV2(address, userVoteList)
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/governance.go:585]: In this function, we do a get -> update -> save -> get (for clear) -> save (after clear). I think we can optimize this by making the `clearUserVotes` do actually one thing: clear the old data for a provided userVoteList and move the call before the last save.

File: vm/systemSmartContracts/governance.go [hunk 2]
```diff
@@ -700,22 +709,77 @@ func (g *governanceContract) closeProposal(args *vmcommon.ContractCallInput) vmc
...
+	}
+
+	g.processLastEndedNonce()
+	return vmcommon.Ok
+}
+
+func (g *governanceContract) clearEndedProposals(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	numAddresses := uint64(len(args.Arguments))
+	if numAddresses == 0 {
+		return vmcommon.Ok
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/governance.go:736]: we do not subtract gas for a call with 0 parameters. Is this ok?

File: vm/systemSmartContracts/governance.go [hunk 3]
```diff
@@ -700,22 +709,77 @@ func (g *governanceContract) closeProposal(args *vmcommon.ContractCallInput) vmc
...
+	}
+
+	lastEndedNonce := g.getLastEndedNonce()
+	for i := uint64(0); i < numAddresses; i++ {
+		voter := args.Arguments[i]
+		if len(voter) != len(args.CallerAddr) {
+			g.eei.AddReturnMessage("invalid delegator address")
+			return vmcommon.UserError
+		}
+		err = g.clearUserVotes(voter, lastEndedNonce)
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/governance.go:752]: from linter: `ineffectual assignment to err (ineffassign)`

File: vm/systemSmartContracts/governance.go [hunk 4]
```diff
@@ -1266,6 +1330,37 @@ func (g *governanceContract) convertV2Config(config config.GovernanceSystemSCCon
 	}, nil
 }
 
+func (g *governanceContract) processLastEndedNonce() {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/governance.go:1333]: how about this code reorg?
```GO
func (g *governanceContract) processLastEndedNonce() {
	lastEndedNonce := g.getLastEndedNonce()
	currentEpoch := g.eei.BlockChainHook().CurrentEpoch()

	maxNoncesToCheck := uint64(1000)
	nonceFound := big.NewInt(int64(lastEndedNonce))
	for nonce := lastEndedNonce + 1; nonce < lastEndedNonce+maxNoncesToCheck; nonce++ {
		nonceBig := big.NewInt(int64(nonce))
		proposal, err := g.getProposalFromNonce(nonceBig)
		if err != nil || proposal.EndVoteEpoch >= uint64(currentEpoch) {
			nonceFound = nonceBig.Sub(nonceBig, big.NewInt(1))
			break
		}
	}

	g.eei.SetStorage([]byte(lastEndedNonceKey), nonceFound.Bytes())
}
```

Comment (iulianpascalau) [reply vm/systemSmartContracts/governance.go:1333]: in this case we have a `maxNoncesToCheck` that will stop the get operations if something goes wrong

File: vm/gasCost.go [hunk 5]
```diff
@@ -28,6 +28,7 @@ type MetaChainSystemSCsCost struct {
 	DelegateVote          uint64
 	RevokeVote            uint64
 	CloseProposal         uint64
+	ClearProposal         uint64
```
Comment (iulianpascalau) [inline vm/gasCost.go:31]: you did not define the cost in the gas schedules toml files. Also, the tests fail because of this empty gas cost.

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: vm/systemSmartContracts/governance.go (hunk 1 above)
Comment (sasurobert) [reply vm/systemSmartContracts/governance.go:585]: yep. no need to save 2 times. optimise this.

File: vm/systemSmartContracts/governance.go (hunk 2 above)
Comment (sasurobert) [reply vm/systemSmartContracts/governance.go:736]: everything should consume gas. this is not ok.

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: vm/systemSmartContracts/governance.go (hunk 1 above)
Comment (dragos-rebegea) [reply vm/systemSmartContracts/governance.go:585]: fixed

File: vm/systemSmartContracts/governance.go (hunk 3 above)
Comment (dragos-rebegea) [reply vm/systemSmartContracts/governance.go:752]: fixed

File: vm/systemSmartContracts/governance.go (hunk 2 above)
Comment (dragos-rebegea) [reply vm/systemSmartContracts/governance.go:736]: fixed

File: vm/systemSmartContracts/governance.go (hunk 4 above)
Comment (dragos-rebegea) [reply vm/systemSmartContracts/governance.go:1333]: 👍

File: vm/gasCost.go (hunk 5 above)
Comment (dragos-rebegea) [reply vm/gasCost.go:31]: fixed

File: cmd/node/config/gasSchedules/gasScheduleV7.toml [hunk 6]
```diff
@@ -41,6 +41,7 @@
     DelegateVote          = 50000000
     RevokeVote            = 50000000
     CloseProposal         = 50000000
+    ClearProposal         = 50000000
```
Comment (iulianpascalau) [inline cmd/node/config/gasSchedules/gasScheduleV7.toml:44]: when you will merge the feat branch, please take into account that we have also the gasScheduleV8.toml

File: vm/systemSmartContracts/governance.go [hunk 7]
```diff
@@ -753,34 +753,40 @@ func (g *governanceContract) clearEndedProposals(args *vmcommon.ContractCallInpu
...
-		Direct:               make([]uint64, 0),
-		DelegatedWithAddress: make([]*DelegatedWithAddress, 0),
-	}
-
-	for _, delegatedWithAddress := range userVoteList.DelegatedWithAddress {
-		if delegatedWithAddress.Nonce > lastEndedNonce {
-			clearedUserVoteList.DelegatedWithAddress = append(clearedUserVoteList.DelegatedWithAddress, delegatedWithAddress)
+func (g *governanceContract) clearUserVotesForList(userVoteList *OngoingVotedListV2, lastEndedNonce uint64) {
+	filteredDirect := make([]uint64, 0)
+	for _, direct := range userVoteList.Direct {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/governance.go:767]: L767-L771 could have been a function that could have been reused also for the delegated list

**Review by iulianpascalau — APPROVED** [review]

File: cmd/node/config/gasSchedules/gasScheduleV7.toml (hunk 6 above)
Comment (dragos-rebegea) [reply cmd/node/config/gasSchedules/gasScheduleV7.toml:44]: 👍

File: vm/systemSmartContracts/governance.go [hunk 8]
```diff
@@ -703,22 +709,86 @@ func (g *governanceContract) closeProposal(args *vmcommon.ContractCallInput) vmc
...
+	}
+
+	g.processLastEndedNonce()
+	return vmcommon.Ok
+}
+
+func (g *governanceContract) clearEndedProposals(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	numAddresses := uint64(len(args.Arguments))
+	perAddressGas := g.gasCost.MetaChainSystemSCsCost.ClearProposal
+	totalGas := numAddresses * perAddressGas
```
Comment (sasurobert) [inline vm/systemSmartContracts/governance.go:736]: (numAddresses + 1) * perAddressGas and you can delete next if.

Comment (dragos-rebegea) [reply vm/systemSmartContracts/governance.go]: done

File: vm/systemSmartContracts/governance.go [hunk 9]
```diff
@@ -708,22 +714,79 @@ func (g *governanceContract) closeProposal(args *vmcommon.ContractCallInput) vmc
...
+	if !g.enableEpochsHandler.IsFlagEnabled(common.GovernanceFixesFlag) {
+		return vmcommon.Ok
+	}
+
+	g.processLastEndedNonce()
+	return vmcommon.Ok
+}
+
+func (g *governanceContract) clearEndedProposals(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	numAddresses := uint64(len(args.Arguments))
```
Comment (sasurobert) [inline vm/systemSmartContracts/governance.go:739]: you need to check that args.CallValue to be equal with 0.

**Review by iulianpascalau — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]
